datafusion-examples/examples/function_factory.rs

```diff
@@ -0,0 +1,232 @@
+        .with_function_factory(Arc::new(CustomFunctionFactory::default()));
+
+    // With the function factory, we can now call `CREATE FUNCTION` SQL functions
+
+    // Let us register a function called f which takes a single argument and
+    // returns that value plus one
+    let sql = r#"
+    CREATE FUNCTION f1(BIGINT)
+        RETURNS BIGINT
+        RETURN $1 + 1
+    "#;
+
+    ctx.sql(sql).await?.show().await?;
+
+    // Now, let us register a function called f2  which takes two arguments and
+    // returns the first argument added to the result of calling f1 on that
+    // argument
+    let sql = r#"
+    CREATE FUNCTION f2(BIGINT, BIGINT)
+        RETURNS BIGINT
+        RETURN $1 + f1($2)
+    "#;
+
+    ctx.sql(sql).await?.show().await?;
+
+    // Invoke f2, and we expect to see 1 + (1 + 2) = 4
+    // Note this function works on columns as well as constants.
+    let sql = r#"
+    SELECT f2(1, 2)
+    "#;
+    ctx.sql(sql).await?.show().await?;
+
+    // Now we clean up the session by dropping the functions
+    ctx.sql("DROP FUNCTION f1").await?.show().await?;
+    ctx.sql("DROP FUNCTION f2").await?.show().await?;
+
+    Ok(())
+}
+
+/// This is our FunctionFactory that is responsible for converting `CREATE
+/// FUNCTION` statements into function instances
+#[derive(Debug, Default)]
+struct CustomFunctionFactory {}
+
+#[async_trait::async_trait]
+impl FunctionFactory for CustomFunctionFactory {
+    /// This function takes the parsed `CREATE FUNCTION` statement and returns
+    /// the function instance.
+    async fn create(
+        &self,
+        _state: &SessionConfig,
+        statement: CreateFunction,
+    ) -> Result<RegisterFunction> {
+        let f: ScalarFunctionWrapper = statement.try_into()?;
+
+        Ok(RegisterFunction::Scalar(Arc::new(ScalarUDF::from(f))))
+    }
+}
+
+/// this function represents the newly created execution engine.
+#[derive(Debug)]
+struct ScalarFunctionWrapper {
+    /// The text of the function body, `$1 + f1($2)` in our example
+    name: String,
+    expr: Expr,
+    signature: Signature,
+    return_type: arrow_schema::DataType,
+}
+
+impl ScalarUDFImpl for ScalarFunctionWrapper {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn signature(&self) -> &datafusion_expr::Signature {
+        &self.signature
+    }
+
+    fn return_type(
+        &self,
+        _arg_types: &[arrow_schema::DataType],
+    ) -> Result<arrow_schema::DataType> {
+        Ok(self.return_type.clone())
+    }
+
+    fn invoke(
+        &self,
+        _args: &[datafusion_expr::ColumnarValue],
+    ) -> Result<datafusion_expr::ColumnarValue> {
+        // Since this function is always simplified to another expression, it
+        // should never actually be invoked
+        internal_err!("This function should not get invoked!")
+    }
+
+    /// The simplify function is called to simply a call such as `f2(2)`. This
```

Comment:
I tried to add some comments inline that explained what the functions were doing with more words, but hopefully for someone who was seeing if this worked for their usecase